UE4Project/Plugins/AirSim/Source/Vehicles/Car/CarPawnSimApi.cpp

```diff
@@ -589,11 +591,51 @@ void CarPawnSimApi::updateKinematics(float dt)
     kinematics_->update();
 }
 
+float computeAngleChange(float prev_rotation_angle_rad, float new_rotation_angle_rad) 
+{
+    return new_rotation_angle_rad + (new_rotation_angle_rad > prev_rotation_angle_rad ? 0.0f : 2.0f * M_PI) - prev_rotation_angle_rad;
+}
+
+void CarPawnSimApi::updateWheelStates(float dt)
+{
+    wheel_states_->time_stamp = msr::airlib::ClockFactory::get()->nowNanos();
+
+    const float unreal_rotation_angle_to_radians = (2 * M_PI) / -1800.0;
+    const float unreal_steering_angle_to_radians = (M_PI) / 180.0;
+    const float radians_per_second_to_rpm = 60.0 / (2 * M_PI);
+
+    //update kinematics from pawn's movement instead of physics engine
+    float fl_new_rotation_angle_rad = movement_->Wheels[0]->GetRotationAngle() * unreal_rotation_angle_to_radians;
+    float fr_new_rotation_angle_rad = movement_->Wheels[1]->GetRotationAngle() * unreal_rotation_angle_to_radians;
+    float rl_new_rotation_angle_rad = movement_->Wheels[2]->GetRotationAngle() * unreal_rotation_angle_to_radians;
+    float rr_new_rotation_angle_rad = movement_->Wheels[3]->GetRotationAngle() * unreal_rotation_angle_to_radians;
+
+    wheel_states_->fl.rpm = 0.3 * wheel_states_->fl.rpm + 0.7 * computeAngleChange(wheel_states_->fl.rotation_angle, fl_new_rotation_angle_rad) / dt * radians_per_second_to_rpm;
```

Comment:
The multiplication of 0.3 and 0.7 is to have some filtering I suppose? In case you have wheelspin?

Reply:
That's right. Empirically, those values smooth out the RPM measurement, while still keeping it close to the instantaneous value